File: 757-set-intersection-size-at-least-two.py
from collections import defaultdict


class Solution:
    def intersectionSizeTwo(self, intervals: list[list[int]]) -> int:
        '''
        Using greedy approach, we first sort the intervals by the end_time.
        We have a set of `nums` that we picked
        Next, we go through each sorted interval, we check how many of the last 2 points in `nums` are part of the current interval.
            - 0: add current's end_time and end_time - 1 into `nums`
            - 1: add current's end_time into `nums`
            - 2: Do nothing

        At the end, return len(nums)
        '''
        nums = []
        sorted_intervals = sorted(intervals, key=lambda i: i[1])
        for start, end in sorted_intervals:
            count = 0
            if nums and nums[-1] >= start:
                count += 1
            if len(nums) >= 2 and nums[-2] >= start:
                count += 1

            if count == 0:
                nums.append(end - 1)
                nums.append(end)
            elif count == 1:
                if nums[-1] == end:
                    nums.append(end - 1)
                else:
                    nums.append(end)

        return len(nums)


sol = Solution()
print(
    f'output: {sol.intersectionSizeTwo([[1, 3], [3, 7], [8, 9]])}, expected: 5')
print(
    f'output: {sol.intersectionSizeTwo([[1, 3], [1, 4], [2, 5], [3, 5]])}, expected: 3')
print(
    f'output: {sol.intersectionSizeTwo([[1, 2], [2, 3], [2, 4], [4, 5]])}, expected: 5')
print(
    f'output: {sol.intersectionSizeTwo([[1, 3], [3, 7], [5, 7], [7, 8]])}, expected: 5')

File: 757-set-intersection-size-at-least-two.java
import java.util.*;

class SetIntersectionSizeAtLeastTwo {

    public int intersectionSizeTwo(int[][] intervals) {
        List<Integer> nums = new ArrayList<>();
        Arrays.sort(intervals, (i1, i2) -> Integer.compare(i1[1], i2[1]));
        for (int[] interval : intervals) {
            int start = interval[0], end = interval[1];
            int count = 0;
            if (nums.size() > 0 && nums.get(nums.size() - 1) >= start) {
                count++;
            }
            if (nums.size() >= 2 && nums.get(nums.size() - 2) >= start) {
                count++;
            }

            if (count == 0) {
                nums.add(end - 1);
                nums.add(end);
            } else if (count == 1) {
                if (nums.get(nums.size() - 1) == end) {
                    nums.add(end - 1);
                } else {
                    nums.add(end);
                }
            }
        }
        return nums.size();
    }

    public static void main(String[] args) {
        SetIntersectionSizeAtLeastTwo sol = new SetIntersectionSizeAtLeastTwo();
        System.out.println(
                "Output: " + sol.intersectionSizeTwo(new int[][] { { 1, 3 }, { 3, 7 }, { 8, 9 } }) + ", expected: 5");
        System.out.println("Output: " + sol.intersectionSizeTwo(new int[][] { { 1, 3 }, { 1, 4 }, { 2, 5 }, { 3, 5 } })
                + ", expected: 3");
        System.out.println("Output: " + sol.intersectionSizeTwo(new int[][] { { 1, 2 }, { 2, 3 }, { 2, 4 }, { 4, 5 } })
                + ", expected: 5");
    }
}
